### DISCUSSION TITLE
I add a waterfall

### DISCUSSION THREAD

**Discussion by RepeatedKibbles** [Techokami/SonicWorldsNext#27] [Show and tell]
https://user-images.githubusercontent.com/101306285/219970949-b8817b78-cc23-4f06-aacd-c4a4873a6743.mp4

I'm not a good explainer for some reason, but I end up in the final for tiles alongside animation.tres (if you know what I'm talking about) but it looks cool, although this tree is behind of waterfall due to difficulties with the engine itself what I'm trying to do.

https://github.com/RepeatedKibbles/SonicWorldsNext

**Comment by DimensionWarped** [comment]
When I said you need to make it tileable, I was referring to the large bulky size of the images. Simply cutting them up and putting them into individual image files doesn't meet the requirements I set for you here.

The waterfall graphics are composed of the same 16x16 tile repeated over the width and height of the waterfall barring the top which is just another line of repeating 16x16 tiles.

In the original game, these two tiles were actually all there was. The moving pattern where the waterfall appears to be flowing is accomplished through cycling the palette. We don't actually have to do it that way though. 16x16 tiles are tiny enough that adding more isn't really an issue.

If you look at Tools/Graphics/AnimatedTextureDrawer.tscn/.gd, you'll find where I made a simple color changing ring into an animated sprite just by setting the region differently during each frame of the animation. If you go with this method, you should have the following images for your waterfall:

waterfall top: 16 pixel wide, 64 pixel tall image with the four frames of the waterfall top graphic spread out vertically. This image should be set up to tile horizontally.

waterfall base: 4 16x16 pixel images that tile both horizontally and vertically. You'll need to pick which one to use to draw with in animation via script *similar* to the one used for the AnimatedTextureDrawer. (Slightly different though because that one doesn't have to draw from multiple textures)

Ideally, your waterfall should then be scripted to do the following:

* Allow us to set how many tiles wide it is
* Allow us to set how many tiles tall it is
* Provide tool scripts for viewing how it looks as a result of those parameters
* Bonus points for if it is animated in the editor.

Your waterfall graphic should automatically be rescaled to fit the specified length and width.

Another potential issue I can see with this is it probably doesn't respect the water level changes in BaseZone Act 1. Consider moving it to BaseZone Act 2 where I won't likely be using waterlevel changes (since that level is intended to demonstrate a different set of mechanics from BaseZone Act 1)

**Reply by DimensionWarped** [reply]
Oh, btw, if you want to get your tree to be above the waterfall you need to move the waterfall's Z-Index to be further behind the default layer. Take note of the 'Z As Relative'' property too.

**Reply by DimensionWarped** [reply]
Actually, an even better thing to look at for how I achieved tiling like what I want you to do here would be the FBZPylon gimmick.

**Reply by DimensionWarped** [reply]
Quick correction, it seems the top piece of the waterfall tiles ever 32 pixels across. Other details are accurate though.

**Comment by DimensionWarped** [comment]
@Renhoex 

Another concern I have with this is that waterfalls in Sonic games rely on dithering to achieve their desired effect, which doesn't really work on LCD monitors. Perhaps we should consider using transparency/shaders and filling in the image instead. Though that's less authentic, so I'm not sure if it fits with the goals of the project.

**Reply by DimensionWarped** [reply]
Checked Sonic Origins and it just uses the normal graphics made for dithering, so I guess we'll go with that approach. I'm going to go ahead and just whip this up real quick...

**Comment by DimensionWarped** [comment]
@RepeatedKibbles You don't appear in the reviewer options, so I'm just going to ask you to take a look at the pull request I added just now to add the waterfall and tell me if it meets your needs.

https://github.com/Techokami/SonicWorldsNext/pull/28

**Reply by RepeatedKibbles** [reply]
I can tell the tree should be in front of the waterfall, not behind the waterfall. (Sorry for not responding and sorry if I sounded mean or weird)
